Source/WebCore/platform/graphics/skia/FloatRoundedRectSkia.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FloatRoundedRect.h"
+
+#if USE(SKIA)
+#include <skia/core/SkRRect.h>
+
+namespace WebCore {
+
+FloatRoundedRect::FloatRoundedRect(const SkRRect& skRect)
+    : m_rect(skRect.rect())
+{
+    SkVector corner = skRect.radii(SkRRect::kUpperLeft_Corner);
+    m_radii.setTopLeft({ corner.x(), corner.y() });
+    corner = skRect.radii(SkRRect::kUpperRight_Corner);
+    m_radii.setTopRight({ corner.x(), corner.y() });
+    corner = skRect.radii(SkRRect::kLowerRight_Corner);
+    m_radii.setBottomRight({ corner.x(), corner.y() });
+    corner = skRect.radii(SkRRect::kLowerLeft_Corner);
+    m_radii.setBottomLeft({ corner.x(), corner.y() });
+}
+
+FloatRoundedRect::operator SkRRect() const
+{
+    if (!isRounded())
+        return SkRRect::MakeRect(rect());
+
+    SkVector radii[4];
+    radii[SkRRect::kUpperLeft_Corner].set(m_radii.topLeft().width(), m_radii.topLeft().height());
+    radii[SkRRect::kUpperRight_Corner].set(m_radii.topRight().width(), m_radii.topRight().height());
+    radii[SkRRect::kLowerRight_Corner].set(m_radii.bottomRight().width(), m_radii.bottomRight().height());
+    radii[SkRRect::kLowerLeft_Corner].set(m_radii.bottomLeft().width(), m_radii.bottomLeft().height());
```

Comment:
Any reason to not use an initializer expression? The code removed from `GraphicsContextSkia.cpp` did this:

```suggestion
    SkVector radii[4] = {
        {m_radii.topLeft().width(), m_radii.topLeft().height()},
        {m_radii.topRight().width(), m_radii.topRight().height()},
        {m_radii.bottomRight().width(), m_radii.bottomRight().height()},
        {m_radii.bottomLeft().width(), m_radii.bottomLeft().height()},
    };
```

Reply:
I'm not Carlos, but this opaque vector constructor was probably part of the reason why the borders are swapped and nobody noticed. I wonder if this would be an acceptable approach (using array indexing in the initializer):

```cpp
    SkVector radii[4] = {
        [SkRRect::kUpperLeft_Corner] = { m_radii.topLeft().width(), m_radii.topLeft().height() },
        [SkRRect::kUpperRight_Corner] = { m_radii.topRight().width(), m_radii.topRight().height() },
        [SkRRect::kLowerRight_Corner] = { m_radii.bottomRight().width(), m_radii.bottomRight().height() },
        [SkRRect::kLowerLeft_Corner] = { m_radii.bottomLeft().width(), m_radii.bottomLeft().height() },
    };
```

Reply:
Yes, I explained it in the other comment, with that initialization is easy to make a mistak because you need to know 0 is top left, 1 is to right, etc. And we actually made that mistake and corners are swapped. I prefer this way explicitly using the Corner enum values.

Reply:
Right, all fine then.

Also, I think the `{ [index] = value }` syntax is a compiler extension. Not sure if it's covered by the designated initializers feature but that would be C++20&mdash;and we are not yet requiring a C++20 compiler anyway.